lib/src/embedded/dispatcher.ts

```diff
@@ -0,0 +1,286 @@
+    return this.onOutboundRequest(
+      handler,
+      OutboundMessage.MessageCase.FUNCTIONCALLREQUEST,
+      InboundMessage.MessageCase.FUNCTIONCALLRESPONSE
+    );
+  }
+
+  // Sends a `request` of type `requestType` inbound. Returns a promise that
+  // resolves with an outbound response of type `responseType` answers the
+  // request, or errors if any errors were encountered.
+  private handleInboundRequest(
+    request: InboundRequest,
+    requestType: InboundRequestType,
+    responseType: OutboundResponseType
+  ): Promise<OutboundResponse> {
+    request.setId(this.pendingInboundRequests.nextId);
+    // Avoid the race condition in which the response arrives before we even
+    // start listening for it.
+    process.nextTick(() => this.sendInboundMessage(request, requestType));
+
+    return new Promise((resolve, reject) => {
+      this.messages$
+        .pipe(
+          filter(message => message.type === responseType),
+          map(message => message.payload as OutboundResponse),
+          filter(response => response.getId() === request.getId())
+        )
+        .subscribe(response => resolve(response));
+
+      this.error$.subscribe(error => reject(error));
+    });
+  }
+
+  // Subscribes to outbound requests of type `requestType`, registering a
+  // callback that runs when a request arrives. Sends the result of running the
+  // callback (which is of type `responseType`) inbound. Returns the
+  // subscription so that consumers can unsubscribe.
+  private onOutboundRequest<
+    T1 extends OutboundRequest,
+    T2 extends InboundResponse
+  >(
+    handler: (request: T1) => PromiseOr<T2>,
+    requestType: OutboundRequestType,
+    responseType: InboundResponseType
+  ): Subscription {
+    return this.messages$
+      .pipe(
+        filter(message => message.type === requestType),
+        map(message => message.payload as OutboundRequest)
+      )
+      .subscribe(async request => {
+        try {
+          const response = await handler(request as T1);
+          response.setId(request.getId());
+          this.sendInboundMessage(response, responseType);
+        } catch (error) {
+          this.error$.next(error);
+        }
+      });
+  }
+
+  // Sends a message inbound. Keeps track of all pending inbound requests.
+  private sendInboundMessage(
+    payload: InboundRequest | InboundResponse,
+    type: InboundRequestType | InboundResponseType
+  ): void {
+    if (type === InboundMessage.MessageCase.COMPILEREQUEST) {
+      this.pendingInboundRequests.add(
+        payload.getId(),
+        OutboundMessage.MessageCase.COMPILERESPONSE
+      );
+    } else if (
+      type === InboundMessage.MessageCase.IMPORTRESPONSE ||
+      type === InboundMessage.MessageCase.FILEIMPORTRESPONSE ||
+      type === InboundMessage.MessageCase.CANONICALIZERESPONSE ||
+      type === InboundMessage.MessageCase.FUNCTIONCALLRESPONSE
+    ) {
+      this.pendingOutboundRequests.resolve(payload.getId(), type);
+    }
```

Comment:
Is there another possible `type`? Would it make sense to add an assertion to that effect so that if one gets added in the future it throws an error that's easy to detect?